### PR TITLE
Feature: Add JUnit 5 support

### PR DESCRIPTION
**Pull request by mojtab23**

### CODE REVIEW

File: src/main/java/ir/sahab/zookeeperrule/ZooKeeperBase.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+     * caller is responsible to close the returned client.
+     */
+    public CuratorFramework newClient() {
+        return newClient(null);
+    }
+
+    /**
+     * Returns a new initialized client which has address of this embedded ZK as its remote and is
+     * working in the specified namespace. The caller is responsible to close the returned client.
+     * The caller is responsible to close it.
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/zookeeperrule/ZooKeeperBase.java:82]: It is duplicate. please remove it.

File: src/main/java/ir/sahab/zookeeperrule/ZooKeeperBase.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+import org.apache.curator.framework.CuratorFrameworkFactory;
+import org.apache.curator.retry.RetryNTimes;
+import org.apache.curator.test.TestingServer;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.nio.file.Files;
+
+class ZooKeeperBase {
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/zookeeperrule/ZooKeeperBase.java]: Add java doc for class

File: src/test/java/ir/sahab/zookeeperrule/ZooKeeperExtensionTest.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    private void checkWriteAndRead(CuratorFramework client, String path)
+            throws Exception {
+        byte[] zNodeDataBytes = "data".getBytes();
+        client.create().forPath(path, zNodeDataBytes);
+        byte[] resultBytes = client.getData().forPath(path);
+        assertArrayEquals(resultBytes, zNodeDataBytes);
+    }
+
+    private void checkZkIsClean() throws Exception {
+        CuratorFramework client = zooKeeperExtension.newClient();
```
Comment (saeidrastak) [inline src/test/java/ir/sahab/zookeeperrule/ZooKeeperExtensionTest.java]: Use try with resources to close the client.

File: src/test/java/ir/sahab/zookeeperrule/ZooKeeperExtensionTest.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        List<String> children = client.getChildren().forPath("/");
+
+        // By default, a node named "zookeeper" exists under root.
+        // So we expect here the size 1 instead of size 0.
+        assertEquals(1, children.size());
+        assertEquals("zookeeper", children.get(0));
+    }
+
+    private void makeZkDirty()
+            throws Exception {
```
Comment (saeidrastak) [inline src/test/java/ir/sahab/zookeeperrule/ZooKeeperExtensionTest.java]: move the throws part one line up

File: src/test/java/ir/sahab/zookeeperrule/ZooKeeperExtensionTest.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    void testExplicitAddress() throws Exception {
+        checkZkIsClean();
+
+        assertEquals(LOCAL_PORT, zkClassWideExtension.getPort());
+
+        makeZkDirty();
+    }
+
+    private void checkWriteAndRead(CuratorFramework client, String path)
+            throws Exception {
```
Comment (saeidrastak) [inline src/test/java/ir/sahab/zookeeperrule/ZooKeeperExtensionTest.java]: move the throws part one line up

File: src/test/java/ir/sahab/zookeeperrule/ZooKeeperExtensionTest.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        // Check client without namespace
+        try (CuratorFramework client = zkClassWideExtension.newClient()) {
+            checkWriteAndRead(client, "/no-namespace-path");
+        }
+
+        // Check client with specific namespace
+        try (CuratorFramework client = zkClassWideExtension.newClient("test-ns")) {
+            checkWriteAndRead(client, "/namespace-path");
+        }
+        // Check external client
```
Comment (saeidrastak) [inline src/test/java/ir/sahab/zookeeperrule/ZooKeeperExtensionTest.java:40]: Add a line break here

File: src/test/java/ir/sahab/zookeeperrule/ZooKeeperExtensionTest.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        // Check client without namespace
+        try (CuratorFramework client = zooKeeperExtension.newClient()) {
+            checkWriteAndRead(client, "/no-namespace-path");
+        }
+
+        // Check client with specific namespace
+        try (CuratorFramework client = zooKeeperExtension.newClient("test-ns")) {
+            checkWriteAndRead(client, "/namespace-path");
+        }
+        // Check external client
```
Comment (saeidrastak) [inline src/test/java/ir/sahab/zookeeperrule/ZooKeeperExtensionTest.java:64]: Add a line break before this line

File: src/main/java/ir/sahab/zookeeperrule/ZooKeeperBase.java (hunk 1 above)
Comment (mojtab23) [reply src/main/java/ir/sahab/zookeeperrule/ZooKeeperBase.java:82]: I can't, we need this for ZookepperExtension users in order to see the doc.

File: src/main/java/ir/sahab/zookeeperrule/ZooKeeperBase.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+    void setup() throws Exception {
+
+        // ZooKeeperServer overrides DefaultUncaughtExceptionHandler,
+        // and we do not want anyone to override this behaviour.
+        // So here, we are going to back up the DefaultUncaughtExceptionHandler before
+        // creating ZkServer and restore it after.
+        Thread.UncaughtExceptionHandler handler = Thread.getDefaultUncaughtExceptionHandler();
+
+        File tempDirectory = Files.createTempDirectory("zk-test").toFile();
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/zookeeperrule/ZooKeeperBase.java:46]: Suppress java:S5443 sonar warning

File: src/test/java/ir/sahab/zookeeperrule/ZooKeeperExtensionTest.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    void testExplicitAddress() throws Exception {
+        checkZkIsClean();
+
+        assertEquals(LOCAL_PORT, zkClassWideExtension.getPort());
+
+        makeZkDirty();
+    }
+
+    private void checkWriteAndRead(CuratorFramework client, String path)
+            throws Exception {
```
Comment (saeidrastak) [inline src/test/java/ir/sahab/zookeeperrule/ZooKeeperExtensionTest.java]: move the throws part one line up

**Pull request merged**